File: domain/src/main/scala/net/kemitix/thorp/domain/RemoteObjects.scala
package net.kemitix.thorp.domain

import zio.UIO

import scala.collection.MapView

/**
  * A list of objects and their MD5 hash values.
  */
final case class RemoteObjects private (
    byHash: MapView[MD5Hash, RemoteKey],
    byKey: MapView[RemoteKey, MD5Hash]
)

object RemoteObjects {

  val empty: RemoteObjects = RemoteObjects(MapView.empty, MapView.empty)

  def create(byHash: MapView[MD5Hash, RemoteKey],
             byKey: MapView[RemoteKey, MD5Hash]): RemoteObjects =
    RemoteObjects(byHash, byKey)

  def remoteKeyExists(
      remoteObjects: RemoteObjects,
      remoteKey: RemoteKey
  ): UIO[Boolean] = UIO(remoteObjects.byKey.contains(remoteKey))

  def remoteMatchesLocalFile(
      remoteObjects: RemoteObjects,
      localFile: LocalFile
  ): UIO[Boolean] =
    UIO(
      remoteObjects.byKey
        .get(localFile.remoteKey)
        .exists(LocalFile.matchesHash(localFile)))

  def remoteHasHash(
      remoteObjects: RemoteObjects,
      hashes: Hashes
  ): UIO[Option[(RemoteKey, MD5Hash)]] =
    UIO(remoteObjects.byHash.collectFirst {
      case (hash, key) if (hashes.values.exists(h => h == hash)) => (key, hash)
    })

}

File: domain/src/main/java/net/kemitix/thorp/domain/RemoteObjects.java
package net.kemitix.thorp.domain;

import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;

import java.util.Map;
import java.util.Optional;

@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
public class RemoteObjects {
    public final MapView<MD5Hash, RemoteKey> byHash;
    public final MapView<RemoteKey, MD5Hash> byKey;
    public static final RemoteObjects empty =
            new RemoteObjects(MapView.empty(), MapView.empty());
    public static RemoteObjects create(
            Map<MD5Hash, RemoteKey> byHash,
            Map<RemoteKey, MD5Hash> byKey
    ) {
        return new RemoteObjects(
                MapView.of(byHash),
                MapView.of(byKey)
        );
    }
    public boolean remoteKeyExists(RemoteKey remoteKey) {
        return byKey.contains(remoteKey);
    }
    public boolean remoteMatchesLocalFile(LocalFile localFile) {
        return byKey.get(localFile.remoteKey)
                .map(localFile::matchesHash)
                .orElse(false);
    }
    public Optional<Tuple<RemoteKey, MD5Hash>> remoteHasHash(Hashes hashes) {
        return byHash.collectFirst(
                (hash, key) -> hashes.values().contains(hash))
                .map(Tuple::swap);
    }
}
